File: 0101-0200/0198-house-robber.rs
struct Solution {}
struct Solution2 {}

impl Solution2 {
    pub fn rob(nums: Vec<i32>) -> i32 {
        if nums.is_empty() {
            return 0;
        }
        let mut rob = nums[0];
        let mut norob = 0;
        for x in nums[1..].iter() {
            let old_rob = rob;
            rob = norob + x;
            norob = norob.max(old_rob);
        }
        rob.max(norob)
    }
}

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        if nums.len() <= 2 {
            return *nums.iter().max().unwrap_or(&0); 
        }
        let mut two = nums[0];
        let mut one = nums[0].max(nums[1]);
        for x in nums[2..].iter() {
            let old_one = one;
            // 注意这里
            one = (two + x).max(one);
            two = old_one;
        }
        one.max(two)
    }
}


fn main() {
    assert_eq!(Solution::rob(vec![1, 2, 3, 1]), 4);
    assert_eq!(Solution::rob(vec![2, 7, 9, 3, 1]), 12);
    assert_eq!(Solution::rob(vec![2, 1, 1, 2]), 4);
}

File: 0101-0200/0198-house-robber.py
from typing import List

class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        rob = nums[0]
        norob = 0
        for i in range(1, len(nums)):
            new_norob = max(norob, rob)
            rob = norob + nums[i]
            norob = new_norob
        return max(rob, norob)

# 定义好状态: 是到第i天(不管第i天有没有盗窃)的最大金额
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        if len(nums) <= 2:
            return max(nums)
        two, one = nums[0], max(nums[0:2])
        for i in range(2, len(nums)):
            old_one = one
            # 注意
            one = max(two + nums[i], one)
            two = old_one
        return max(one, two)

print(Solution().rob([1,2,3,1]))
print(Solution().rob([2,7,9,3,1]))
print(Solution().rob([2,1,1,2]))
print(Solution().rob([1,3,1,3,100]))
